_posts/2016-12-7-algorithms4-chapter2.md

```diff
@@ -0,0 +1,261 @@
+	{ return v.compareTo(w) < 0; }
+	 
+	private static void exch(Comparable[] a, int i, int j)
+	{ Comparable t = a[i]; a[i] = a[j]; a[j] = t; }
+	
+	private static void show(Comparable[] a)
+	{ // Print the array, on a single line.
+		for (int i = 0; i < a.length; i++)
+			StdOut.print(a[i] + " ");
+		StdOut.println();
+	}
+
+	public static boolean isSorted(Comparable[] a)
+	{ // Test whether the array entries are in order.
+		for (int i = 1; i < a.length; i++)
+			if (less(a[i], a[i-1])) return false;
+		return true;
+	}
+
+	public static void main(String[] args)
+	{ // Read strings from standard input, sort them, and print.
+		String[] a = In.readStrings();
+		sort(a);
+	 	assert isSorted(a);
+		show(a);
+	}
+}
+
+{% endhighlight %}
+
+Sorting cost model.
+When studying sorting algorithms, we count compares and exchanges. 
+For algorithms that do not use exchanges, we count array accesses.
+
+two elementary sorting methods: selection sort and insertion sort
+a variation of one of them: shellsort
+
+##### Selection Sort
+
+Steps:
+* First, find the smallest item in the array, and exchange it with the first entry.
+* Then, find the next smallest item and exchange it with the second entry.
+* Continue in this way until the entire array is sorted.
+
+Two signature properties:
+* Running time is insensitive to input
+* Data movement is minimal
+
+**Proposition A.**
+If the length of an array is N, we uses $N^2/2$ compares and N exchanges to.
+
+{% highlight java %}
+
+public class Selection
+{
+	public static void sort(Comparable[] a)
+ 	{ // Sort a[] into increasing order.
+ 		int N = a.length; // array length
+ 		for (int i = 0; i < N; i++)
+ 		{ // Exchange a[i] with smallest entry in a[i+1...N).
+ 			int min = i; // index of minimal entr.
+ 			for (int j = i+1; j < N; j++)
+ 				if (less(a[j], a[min])) min = j;
+ 			exch(a, i, min);
+ 		}
+ 	}
+ 	// See page 245 for less(), exch(), isSorted(), and main().
+}
+
+{% endhighlight %}
+
+##### Insertion Sort
+
+Unlike that of selection sort, the running time of insertion sort depends on the initial
+order of the items in the input. Before inserting the current item into the vacated 
+position, we need to make space for the current item by moving larger items one position
+to the right. 
+
+{% highlight java %}
+
+public class Insertion
+{
+	public static void sort(Comparable[] a)
+	{ // Sort a[] into increasing order.
+		int N = a.length;
+		for (int i = 1; i < N; i++)
+		{ // Insert a[i] among a[i-1], a[i-2], a[i-3]... ..
+			for (int j = i; j > 0 && less(a[j], a[j-1]); j--)
+				exch(a, j, j-1);
+		}
+	}
+	// See page 245 for less(), exch(), isSorted(), and main().
+}
+
+{% endhighlight %}
+
+**Proposition B.**
+On the average, insertion sort uses $N^2/4$ compares and $N^2/4$ exchanges to sort a randomly
+ordered array of length N with distinct keys. The worst case is $N^2/2$ compares and $N^2/2$ 
+exchanges and the best case is N-1 compares and 0 exchanges.
```

Comment:
It will be better if you change all the numbers into LaTex format. e.g. wrap N-1 with $